File: _64_OddPeriodSquareRoots.java
import java.util.*;

public class _64_OddPeriodSquareRoots{
    
    public static int getRootPeriod(int n){
        double s = Math.sqrt(n);
        
        // Square roots of perfect squares have a period of 0.
        if (s == Math.floor(s)) return 0;
        
        // sqrt(n) = a_0 + sqrt(n) - a_0
        // sqrt(n) = a_0 + (sqrt(n) - b_0)/c_0
        // a_0 = floor(sqrt(n)), b_0 = a_0, c_0 = 1
        
        // sqrt(n) = a_0 + 1/(c_0/(sqrt(n) - b_0))
        // sqrt(n) = a_0 + 1/((sqrt(n) + b_0)/((n - b_0^2)/c_0))
        //     (sqrt(n) + b_0)/((n - b_0^2)/c_0) = a_1 + (sqrt(n) - b_1)/c_1
        // sqrt(n) = a_0 + 1/(a_1 + (sqrt(n) - b_1)/c_1)
        
        // c_1 = (n - b_0^2) / c_0;
        // a_1 = floor((sqrt(n) + b_0)/c_1)
        // b_1 = a_1*c_1 - b_0
        
        // a_0 + 1/(a_1 + (sqrt(n) - (a_1*c_1 + b_0))/c_1)
        // a_0 + 1/(a_1 - a_1 + (sqrt(n) - a_0)/c_1)
        // a_0 + 1/((sqrt(n) + a_0)/(n - a_0^2))
        // a_0 + (n - a_0^2)/(sqrt(n) + a_0)
        // a_0 + sqrt(n) - a_0
        // sqrt(n)
        
        ArrayList<List<Integer>> seen = new ArrayList<>();
        int count = 0;
        
        int a = (int) Math.floor(s);
        int b = a;
        int c = 1;
        
        while (true){
            c = (n - b*b) / c;
            a = (int) Math.floor((Math.sqrt(n) + b)/c);
            b = a*c - b;
            // System.out.println(b + " " + c);
            for (int i = 0; i < seen.size(); i++){
                List<Integer> list = seen.get(i);
                if (list.get(0) == b && list.get(1) == c){
                    // System.out.println(seen);
                    return count - i;
                }
            }
            seen.add(Arrays.asList(new Integer[]{b, c}));
            count++;
        }
    }
    
    public static void main(String[] args){
        
        int numFractions = 0;
        
        for (int n = 2; n <= 10000; n++){
            int period = getRootPeriod(n);
            // System.out.println(n + ": " + period);
            
            if (period % 2 == 1){
                numFractions++;
            }
        }
        
        System.out.println("Number of continued fractions with odd periods: " + numFractions); // 1322
    }
}

File: c#/_64_OddPeriodSquareRoots.cs
using System;
using System.Collections.Generic;

public class _64_OddPeriodSquareRoots{
	public static int Euler64(){
		int res = 0;
		for (int n = 2; n <= 10_000; n++){
			if (GetPeriod(n) % 2 == 1){
				res++;
			}
		}
		return res;
	}
	
	private static int GetPeriod(int n){
		double s = Math.Sqrt(n);
		// if n is a perfect square, period is 0
		if (s == (int) s) return 0;
		
		// sqrt(n) = a + sqrt(n) - a
		// sqrt(n) = a + (sqrt(n) - b) / c
		int a = (int) Math.Floor(s);
		int b = a;
		int c = 1;
		
		// sqrt(n) = a0 + (sqrt(n) - b0) / c0
		// sqrt(n) = a0 + 1 / (c / (sqrt(n) - b0))
		// sqrt(n) = a0 + 1 / ((c * (sqrt(n) + b0)) / (n - b0^2))
		// sqrt(n) = a0 + 1 / ((sqrt(n) + b0) / ((n - b0^2) / c0))
		// sqrt(n) = a0 + 1 / d
		
		// d = (sqrt(n) + b0) / ((n - b0^2) / c0)
		// c1 = (n - b0^2) / c0
		// a1 = floor((sqrt(n) + b0) / c1)
		// b1 = a1 * c1 - b0
		
		Dictionary<(int, int), int> seen = new Dictionary<(int, int), int>();
		int count = 0;
		while (true){
			c = (n - b * b) / c;
			a = (int) Math.Floor((s + b) / c);
			b = a * c - b;
			var key = (b, c);
			if (seen.ContainsKey(key)){
				return count - seen[key];
			}
			seen.Add(key, count++);
		}
	}
	
	public static void Main(){
		Console.WriteLine(Euler64());
	}
}
